File: Leetcode/1306.cpp
class Solution {
public:
    bool has_ans = false;
    vector<bool> vis;
    void go(vector<int> &arr, int start) {
        try {
            if (arr.at(start) == 0) {
                has_ans = true;
                return;
            }
            if (vis.at(start)) {
                return;
            }
            vis.at(start) = true;
            go(arr, start + arr.at(start));
            go(arr, start - arr.at(start));
        } catch (...) { }
    }
    bool canReach(vector<int>& arr, int start) {
        vis.resize(arr.size(), false);
        go(arr, start);
        return has_ans;
    }
};

File: Leetcode/1306/1306.go
var (
	vis []bool
	n   int
	ans bool
)

func dfs(pos int, arr []int) {
	if ans {
		return
	}
	if arr[pos] == 0 {
		ans = true
		return
	}
	vis[pos] = true
	newPos := pos + arr[pos]
	if newPos >= 0 && newPos < n && !vis[newPos] {
		dfs(newPos, arr)
	}
	newPos = pos - arr[pos]
	if newPos >= 0 && newPos < n && !vis[newPos] {
		dfs(newPos, arr)
	}
}

func canReach(arr []int, start int) bool {
	n = len(arr)
	vis = make([]bool, n)
	ans = false
	dfs(start, arr)
	return ans
}